### PR TITLE
support embedded schema for cassandra and sql update-schema

### PR DESCRIPTION
**Pull request by carlydf**
## What changed?

This change adds a flag to `temporal-sql-tool` and `temporal-cassandra-tool`'s `update-schema` and `setup-schema` to specify an embedded schema instead of a path to the schema directory or schema file.

```
NAME:
   temporal-cassandra-tool update-schema - update cassandra schema to a specific version

USAGE:
   temporal-cassandra-tool update-schema [command options] [arguments...]

OPTIONS:
   --version value, -v value      target version for the schema update, defaults to latest
   --schema-dir value, -d value   path to directory containing versioned schema
   --schema-name value, -s value  name of embedded versioned schema, one of: [cassandra/temporal cassandra/visibility]
```

```
NAME:
   temporal-sql-tool update-schema - update sql schema to a specific version

USAGE:
   temporal-sql-tool update-schema [command options] [arguments...]

OPTIONS:
   --version value, -v value      target version for the schema update, defaults to latest
   --schema-dir value, -d value   path to directory containing versioned schema
   --schema-name value, -s value  name of embedded versioned schema, one of: [mysql/v57/temporal mysql/v57/visibility mysql/v8/temporal mysql/v8/visibility postgresql/v96/temporal postgresql/v96/visibility postgresql/v12/temporal postgresql/v12/visibility]
```

## Why?
We want to embed the db schemas in the `temporal-sql-tool` and `temporal-cassandra-tool` with `go:embed` so that customers can simply specify which schema they want using a flag instead of having to copy the schema files to wherever they're running the db tool and passing the path to the schema.

https://github.com/temporalio/temporal/issues/2059

## How did you test it?
I wrote a new unit test for reading the embedded file system, and the efs==nil codepath of the changed functions are all tested by the existing unit tests. I also updated the unit test for config validation to reflect the new changes.

I tested this locally with a locally running Postgres. Since the more complex changes are in `updatetask.go`, which is shared by the sql and the cassandra tool, I think there is no need to also test with a real cassandra db. I also built the cassandra command line tool and confirmed that the flags behaved as expected.

## Potential risks
Schema updates could fail. But the changes are pretty contained to the new `-s` flag which isn't used anywhere in production yet, so the risk is very low. The code for the `-d` flag remained basically the same.

## Is hotfix candidate?
No

### CODE REVIEW

File: tools/common/schema/updatetask.go [hunk 1]
```diff
@@ -392,8 +428,22 @@ func sortAndFilterVersions(versions []string, startVerExcl string, endVerIncl st
...
+	return readSchemaDirEFS(dir, startVer, endVer, logger, nil)
+}
 
-	subDirs, err := os.ReadDir(dir)
+// readSchemaDirEFS returns a sorted list of subdir names that hold
+// the schema changes for versions in the range startVer < ver <= endVer
+// when endVer is empty this method returns all subdir names that are greater than startVer.
+// Compatible with an embedded file system generated by go:embed, calling efs.ReadDir
+// but if efs == nil it uses os.ReadDir
+func readSchemaDirEFS(dir string, startVer string, endVer string, logger log.Logger, efs *embed.FS) ([]string, error) {
```
Comment (tdeebswihart) [inline tools/common/schema/updatetask.go:439]: You could simplify this by taking in an `fs.FS` instead of an `embed.FS`. You can open the "real" filesystem dir as an `FS` by using `os.DirFS(dir)` which would let you unify the code paths

Comment (tdeebswihart) [reply tools/common/schema/updatetask.go:439]: You may need to take more than just `fs.FS` to do everything you want (I didn't dig through all the calls you made), but `fs.ReadFileFS` and `fs.ReadDirFS` look to have what you need

**Comment by tdeebswihart** [comment]
Do the `setup-schema` commands need to be updated as well? Seems like they should be

File: tools/common/schema/handler.go [hunk 2]
```diff
@@ -92,8 +96,20 @@ func validateSetupConfig(config *SetupConfig) error {
...
+	}
+	if len(config.SchemaDir) > 0 && len(config.SchemaName) > 0 {
+		return NewConfigError("either" + flag(CLIOptSchemaDir) + " or " +
+			flag(CLIOptSchemaName) + " must be specified")
+	}
+	if len(config.SchemaName) > 0 {
+		if !slices.Contains(ValidSchemaNames["sql"], config.SchemaName) &&
+			!slices.Contains(ValidSchemaNames["cassandra"], config.SchemaName) {
+			return NewConfigError(flag(CLIOptSchemaDir) + " must be one of: " +
+				fmt.Sprintf("%v or %v", ValidSchemaNames["sql"], ValidSchemaNames["cassandra"]))
```
Comment (tdeebswihart) [inline tools/common/schema/handler.go:111]: Can you do this separately based on whether they'd chosen SQL or Cassandra? That way they're only given the schemas relevant for the tool at hand. Otherwise our `sql` tool print out schemas that should only be available for cassandra (and vice-versa)

File: tools/common/schema/updatetask.go (hunk 1 above)
Comment (carlydf) [reply tools/common/schema/updatetask.go:439]: i did consider using an `fs.FS` + `os.DirFS(dir)` and actually wrote a bunch of code doing that that I then deleted it lol. I absolutely agree the code would be cleaner (no `if efs != nil`) if I did that, but I wasn't sure what to pass in as the `dir` for `os.DirFS` that would be sure to not break any customer's existing filesystem structures for where they are using this tool and storing their schemas.

I chose this route because I felt like it was less risky, given that I can't test for all the edge cases that might exist in our customer's file systems. 

this https://github.com/golang/go/issues/44279 made me concerned about `os.DirFS`, in addition to the note here https://pkg.go.dev/os#DirFS about limitations. if you have a sec to look at `os.DirFS` with me and talk through it together i would be super down to consider!

File: tools/common/schema/handler.go (hunk 2 above)
Comment (carlydf) [reply tools/common/schema/handler.go:111]: i could i just didnt want to "overengineer"..

**Comment by carlydf** [comment]
> Do the `setup-schema` commands need to be updated as well? Seems like they should be

thanks for catching! i missed this cuz the schema path is not required for that command, but you're totally right, i will go add :)

File: tools/common/schema/updatetask.go (hunk 1 above)
Comment (tdeebswihart) [reply tools/common/schema/updatetask.go:439]: Did you try it out on Windows and encounter that bug? It looks like the problem happens when using the filesystem's root. I wonder if it'll have an issue if, say, a user were to specify a path to a different filesystem like `D:\\Foo\Bar\Baz\schemas`.

I'd definitely prefer simpler code; this is an administrator tool and I highly doubt we support schemas that aren't our official ones, so the common usage will just be `temporal-cassandra-tool update-schemas relative-path-to-schema`.

File: tools/common/schema/handler.go (hunk 2 above)
Comment (carlydf) [reply tools/common/schema/handler.go:111]: thinking about this more, I feel like the added messiness in our code to pass through the db type (we would have to kind of break the "common" db schema abstraction) is not worth the benefit to the user, since I think they will be able to figure out from the error message that the "cassandra" options aren't valid for their config if they're using sql, or the other way around.

File: tools/common/schema/setuptask.go [hunk 3]
```diff
@@ -71,16 +75,27 @@ func (task *SetupTask) Run() error {
...
+			if err2 != nil {
+				return fmt.Errorf("error reading file %s: %w", schemaFilePath, err)
+			}
+			stmts, err = persistence.LoadAndSplitQueryFromReaders([]io.Reader{bytes.NewBuffer(schemaBuf)})
+		} else {
+			filePath, err2 := filepath.Abs(config.SchemaFilePath)
+			if err2 != nil {
+				return err
+			}
+			stmts, err = persistence.LoadAndSplitQuery([]string{filePath})
```
Comment (carlydf) [inline tools/common/schema/setuptask.go:94]: I didn't change this to the `os.DirFS(dirPath)` method, because I didn't want to assume that SchemaFilePath had any directory levels, whereas for `SchemaDir` we know it's a directory

File: tools/common/schema/updatetask.go (hunk 1 above)
Comment (carlydf) [reply tools/common/schema/updatetask.go]: got it working with fs.FS

File: schema/embed.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+// Assets returns a file system with the contents of the schema directory
+func Assets() fs.FS {
+	return assets
+}
+
+// Paths returns a list of paths to directories within the schema subdirectory that have versioned schemas in them
+func Paths(dbSubDir string) []string {
+	efs := Assets()
+	dirs := make([]string, 0)
+	_ = fs.WalkDir(efs, dbSubDir, func(path string, d fs.DirEntry, err error) error {
```
Comment (tdeebswihart) [inline schema/embed.go:44]: You should return this error from the function as well so we don't fail silently

File: tools/common/schema/handler.go (hunk 2 above)
Comment (tdeebswihart) [reply tools/common/schema/handler.go]: Our self-hosting users use this tools and I think its worth it to remove a potential source of confusion, but I wont hold this PR up over it.

I think this can be done simply by adding `Name() string` function to the `DB` interface used by this handler code. If you do that we could ask the provided DB object what it is when checking our configuration and only return the schema names that matter.

It'd be maybe 10 lines of code overall:
- The additional `Name` method definition
- ~3 lines in each of `tools/cassandra/cqlclient.go` and `tools/sql/conn.go` to add the `Name` implementations
- A line or two to get the DB's name and check it when validating configuration

File: schema/embed.go (hunk 4 above)
Comment (carlydf) [reply schema/embed.go]: done

File: schema/embed.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+		if err != nil {
+			return err
+		}
+		if d.IsDir() && d.Name() != path {
```
[the remark this answers is not in the document]
Comment (carlydf) [reply schema/embed.go]: I think that was leftover from when I was misunderstanding WalkFunc and had different logic in the if. Tested and removed

File: tools/common/schema/handler.go [hunk 6]
```diff
@@ -41,46 +46,60 @@ func Setup(cli *cli.Context, db DB, logger log.Logger) error {
...
+	if len(config.SchemaFilePath) > 0 && len(config.SchemaName) > 0 {
+		return NewConfigError("either" + flag(CLIOptSchemaFile) + " or " +
+			flag(CLIOptSchemaName) + " must be specified")
+	}
+	if len(config.SchemaName) > 0 {
+		if !slices.Contains(dbschemas.Paths("mysql"), config.SchemaName) &&
+			!slices.Contains(dbschemas.Paths("postgresql"), config.SchemaName) &&
+			!slices.Contains(dbschemas.Paths("cassandra"), config.SchemaName) {
+			return NewConfigError(fmt.Sprintf("%s must be one of: %v",
+				flag(CLIOptSchemaName), validSchemaNames(db.Name())))
```
Comment (tdeebswihart) [inline tools/common/schema/handler.go:100]: Nit: you use `db.Name` here to only check paths for the requested name

File: tools/common/schema/handler.go [hunk 7]
```diff
@@ -108,3 +140,17 @@ func validateUpdateConfig(config *UpdateConfig) error {
...
+	}
+	return ".sql"
+}
+
+func validSchemaNames(dbName string) []string {
+	if dbName == "sql" {
+		return append(dbschemas.Paths("mysql"), dbschemas.Paths("postgresql")...)
+	}
+	return dbschemas.Paths(dbName)
+}
```
Comment (tdeebswihart) [inline tools/common/schema/handler.go:156]: Nit: this feels like something only the db implementation should care about

**Review by tdeebswihart — APPROVED** [review]
Thanks for working with me on some of this cleanup even though it wasn't from your work in the first place. The tool will be in a much better state after this!

File: tools/common/schema/handler.go (hunk 7 above)
Comment (carlydf) [reply tools/common/schema/handler.go:156]: to me it feels like the concept of a "schema name" is specific to the command line tool, since the concept of a "valid schema name" only exists because it's an option of the command line tool. Like, if the DB implementations are used elsewhere in our codebase, the function "db.ValidSchemaNames" would be irrelevant i think? and under that line of thinking, this function is something that the cli db tool is primarily concerned with, and putting it in the db implementation would pollute it with something that is only relevant because of the format of a specific command line option

Comment (tdeebswihart) [reply tools/common/schema/handler.go]: I'm fine with your approach, I was just actually curious if they were used elsewhere
